src/model_signing/_manifest.py

```diff
@@ -0,0 +1,265 @@
+their hashes. When signing the model we generate a manifest from serializing the
+model (that is, computing the hashes for all the objects, according to the
+specific serialization method used).
+
+When verifying the integrity of the model, we extract a manifest from the
+signature, after verifying the authenticity of the signature. Then, we serialize
+the local model and compare the two manifests.
+
+Comparing the manifests can be done by checking that everything matches, or, we
+can only check partial object match. This is useful, for example, for the cases
+where the original model contained files for multiple ML frameworks, but the
+user only uses the model with one framework. This way, the user can verify the
+integrity only for the files that are actually used.
+"""
+
+import abc
+from collections.abc import Iterable, Iterator
+import dataclasses
+import pathlib
+import sys
+
+from typing_extensions import override
+
+from model_signing._hashing import hashing
+
+
+if sys.version_info >= (3, 11):
+    from typing import Self
+else:
+    from typing_extensions import Self
+
+
+@dataclasses.dataclass(frozen=True, order=True)
+class ResourceDescriptor:
+    """A description of any content from any `Manifest`.
+
+    We aim this to be similar to in-toto's `ResourceDescriptor`. To support
+    cases where in-toto cannot be directly used, we make this a dataclass that
+    can be mapped to in-toto when needed, and used as its own otherwise.
+
+    Not all fields from in-toto are specified at this moment. All fields here
+    must be present, unlike in-toto, where all are optional.
+
+    See github.com/in-toto/attestation/blob/main/spec/v1/resource_descriptor.md
+    for the in-toto specification.
+
+    Attributes:
+        identifier: A string that uniquely identifies this `ResourceDescriptor`
+          within the manifest. Depending on serialized format, users might
+          require the identifier to be unique across all manifests stored in a
+          system. Producers and consumers can agree on additional requirements
+          (e.g., several descriptors must have a common pattern for the
+          identifier and the integrity of the model implies integrity of all
+          these items, ignoring any other descriptor). Corresponds to `name`,
+          `uri`, or `content` in in-toto specification.
+        digest: One digest for the item. Note that unlike in-toto, we only have
+          one digest for the item and it is always required.
+    """
+
+    identifier: str
+    digest: hashing.Digest
+
+
+class ManifestKey(metaclass=abc.ABCMeta):
+    """An object that can be a key for an item in a manifest.
+
+    We need to be able to convert the key to string when we serialize the
+    manifest and to rebuild the object from the serialized representation when
+    we deserialize the manifest from the signature.
+    """
+
+    @classmethod
+    @abc.abstractmethod
+    def from_str(cls, s: str) -> Self:
```

Comment:
Just curious, is there anyway to guarantee the subclass also defines a `__str__`?

Reply:
I think that's implicitly declared for all classes. If you don't define it, `__str__` calls `__repr__` and if you don't define that then there's an implicit method. We cannot force a class to redefine, the only thing we can do is to test these via goldens/other tests.